### ISSUE TITLE
Github actions for pylint available in github actions marketplace

### ISSUE THREAD

**Issue by Pierre-Sassoulas** [pylint-dev/pylint#4322]
Github actions can be defined from the repository. GitHub Action allows repo maintainers to quickly add automated pylint testing on all pull requests submitted to their repos. This will focus development efforts on a single, community-supported GitHub Action for pylint instead of a multitude of homegrown solutions with travis or custom github actions for each repository.

**Comment by iamtodor** [comment]
Hello @Pierre-Sassoulas 
Any updates on this issue? 
We are really missing Github actions for pylint

**Comment by Pierre-Sassoulas** [comment]
Hi, no one worked on it yet, contributions are welcome :)

**Comment by DanielNoord** [comment]
I'm working on this in https://github.com/DanielNoord/pylint-action.

I hope to have a first real version ready in the coming weeks and then move it to this repository/organization.

**Comment by jacobtylerwalls** [comment]
@DanielNoord thanks for stepping up. I had a quick skim of the WIP, and I suppose I have a stupid question. I thought issue was just asking for a tiny three-line YAML recipe, to help people get the same experience they get on the command line with defaults. Checkout python, pip install requirements, run pylint. Something like that. What was your interpretation, if I may ask?

**Comment by DanielNoord** [comment]
@jacobtylerwalls I'd like to support two functionalities in the same action:
1) Run pylint over your codebase and warn when there are errors
2) Run pylint over a PR or push and annotate the respective errors

I'm still figuring out how the Github API works so it still needs (a lot of) work, but I think especially the second function would be useful. That's what most people looking for this action are actually looking for I think.

**Comment by DanielNoord** [comment]
Just to show the progress I have been making today. Something like this is what I have in mind 😄 

<img width="761" alt="Schermafbeelding 2022-02-16 om 23 29 19" src="https://user-images.githubusercontent.com/13665637/154368293-e2807805-eed9-40a1-b314-f94315fe6794.png">

**Comment by jacobtylerwalls** [comment]
Awesome. I guess I was just surprised (speaking only from ignorance!) that it takes JS to do that -- I would have thought a pylint feature could just output some JSON like we already do, and the YAML step could feed that to the check API.

**Comment by DanielNoord** [comment]
Yeah, I must confess: after a day of trying to understand the various bits and pieces of documentation on GitHub Actions I am not even sure if this is the right approach 😅 
How I understand it is that the checks API is actually meant for GitHub Apps, similar to the pre-commit app. However, parts of the checks API is also included in the actions/normal API. For example, the option to send Annotation requests. It is a bit more limited than the fully checks API but it works. I think JS does work quite well here and the code is not too difficult to understand.

However, one of the issues I'm running into is how to test all of it. Especially since I have little experience with the JS testing frameworks. A lot of the code works by getting data from the runner environment or input variables from the user's `.yaml` file. For example, I'm currently struggling to replicate the `GET` request that would normally compare the commit in a push against the latest commit on `main`. I'd like to only release this when there is a 100% coverage so this will probably take some time to get right.

For now, I hope to release a `0.0.1` soon(ish) which can:
1) Run on push or pull requests
2) Download the pylint version specified by the user
3) Get the files to lint on from the push or pull request
4) Get the arguments and configuration file location from the user
5) Run pylint with the arguments over selected files
6) Annotate and fail the action accordingly

Imo, that would be a good MVP but obviously I'm open to any suggestions!

**Comment by cdce8p** [comment]
> Just to show the progress I have been making today. Something like this is what I have in mind 😄

I would suggest looking into [problem-matchers](https://github.com/actions/toolkit/blob/main/docs/problem-matchers.md). It might not be as customizable, but likely much easier to implement.

This is what Home-Assistant does:
* [ci.yaml -> pylint](https://github.com/home-assistant/core/blob/f89de613d9449f1edd717f00f333dd0546985fb4/.github/workflows/ci.yaml#L617-L619)
* [pylint.json (regex-matcher)](https://github.com/home-assistant/core/blob/f89de613d9449f1edd717f00f333dd0546985fb4/.github/workflows/matchers/pylint.json)

--
If Node is too difficult, Python does also work. Although dependency management for the action itself might be more difficult.

**Comment by Pierre-Sassoulas** [comment]
This is amazing @DanielNoord, the screenshot looks great ! 

Regarding the MVP I think it needs something to define what the dependencies are and install them between 1. and 3, because pylint can have a lot of false positive ``no-member`` / ``no-name-in-module`` when it can't import dependencies.  It's a "hard" problem to solve considering all the existing possible solutions to define dependencies in a project (pyproject.toml, setup.py, setup.cfg, requirements.txt, other non-standard solutions...)

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I think that's something for the user to fix (although I should probably document it). If dependencies need to be installed from `requirements.txt` you just need to add an additional step to the action to download those dependencies. Same goes for the other types of dependencies.

@cdce8p Ah, that actually looks quite good.. Kind of annoying how diverse and dispersed the Github documentation is. I never saw that functionality. That said, I don't think there are any inherent limits to the approach I'm taking now so I might as well finish it. It's a good way to brush up on my JS.

**Comment by skhomuti** [comment]
Just leave it here:
I've released a [Pylint plugin](https://github.com/skhomuti/pylint-actions) with a support GitHub Actions reporting format
Look at this how it can be looking https://github.com/skhomuti/pylint-actions/pull/5/files#diff-151cf6ca7f8405370f79d25d426484db7b94be6139d027568f10f3cebe68f18b

I don't think we should have a separate Action to run pylint (because usually, it's just a simple command-line call and nothing more). That's why reporter as a separate plugin makes sense on my view

**Comment by jzohrab** [comment]
Just for limiting scope, an opinion: I don't feel an action is necessary, as devs likely ought to be linting their code locally, and so will have pylint installed as part of the dev requirements anyway.  Their linting steps for local installs should deal with any project-specific weirdness.  They could reuse that in github.  Perhaps a small doc would be sufficient explaining how it can be done.  e.g., here's a rough setup:

```
  code-quality:
    runs-on: ubuntu-latest
    timeout-minutes: 30

    strategy:
      matrix:
        python_version: [ '3.10', '3.11' ]

    steps:
    - uses: actions/checkout@v4
      with:
        submodules: true

    - uses: actions/setup-python@v4
      with:
        python-version: ${{ matrix.python_version }}
        cache: 'pip' # caching pip dependencies

    - run: |
        pip install flit
        flit install --only-deps --deps develop

    - name: Lint
      run: pylint --msg-template='{path} ({line:03d}): {msg} ({msg_id} {symbol})' lute/ tests/
```

I actually use `invoke`, and so have `inv lint` in github, but that's a separate thing.

(Opinion only, in hopes of managing backlog.  I'm happy to contribute a doc change if requested).

**Comment by webknjaz** [comment]
@DanielNoord

> How I understand it is that the checks API is actually meant for GitHub Apps, similar to the pre-commit app. However, parts of the checks API is also included in the actions/normal API. For example, the option to send Annotation requests. It is a bit more limited than the fully checks API but it works. I think JS does work quite well here and the code is not too difficult to understand.

GHA is built on top of GH Apps. That's why these APIs are available there. You don't technically need JS to interact with said API. There are Python clients to it as well.

Though, GHA has a mechanism called workflow commands. And a few of those translate into what's then displayed as annotations. All you have to do is writing those commands to stdout or stderr: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-error-message.

Note, though, that there are limits on how many annotations you can post. It looks like it's still max 10 per step and max 50 per job and workflow run: https://github.com/orgs/community/discussions/26680#discussioncomment-3252835.

This feature as other limitations like immutability — for example, once an annotation shows up in the PR and is fixed by a follow-up commit, it keeps being displayed. I think this is why Codecov is deprecating the use of this feature in their GH App, for instance.

I was playing with GH APIs a lot back when I was building octomachinery more actively — feel free to ask me about these, I may be able to help quicker than experimentation sometimes :)

Among the libs to look into are gidgethub (octomachinery wraps it) and pygithub.

Here's some examples of not needing JS with composite actions, by the way:
* https://github.com/re-actors/alls-green/blob/223e4bb/action.yml#L51C7-L53C45
* https://github.com/pypa/gh-action-pypi-publish/blob/8cafb5c/action.yml#L152C37-L152C60
* https://github.com/ansible-community/ansible-test-gh-action/blob/6b0ca88/action.yml#L218C5-L218C18
* https://github.com/deadsnakes/action/blob/f3df074/bin/install-python

**Comment by webknjaz** [comment]
> Just for limiting scope, an opinion: I don't feel an action is necessary, as devs likely ought to be linting their code locally, and so will have pylint installed as part of the dev requirements anyway.

To expand on this, I also relate to this point. My typical setup is having tox calling pre-commit that has pylint in it. And so I only need to configure running tox in CI and nothing else special is really needed (short of duplicating the deps list in pre-commit, sadly).

So an alternative could be env detection within pylint and just printing out workflow commands to stderr if it's GHA. This could of course be on a command runner level (tox/nox/etc) but then, it's more work/duplication for the end-users (here's an ugly example: https://github.com/ansible/awx-plugins/blob/53ebc59/tox.ini#L293C9-L293C42).
Back to the pylint runtime, I think it might be reasonable to have a pylint plugin doing this. So the UX for the end-users would be just installing it.

Also, have you considered producing a rich output via job summaries? This might be less invasive and more colorful than annotations.
